avr-hal-generic/src/serial.rs

```diff
@@ -1,5 +1,102 @@
 //! Serial Implementations
 
+use core::cmp::Ordering;
+
+// Clock is needed because the calculations needs to take core clock into account
+#[derive(Debug, Clone, Copy)]
+pub struct Baudrate<CLOCK> {
+    pub ubrr: u16,
+    pub u2x: bool,
+    pub _clock: ::core::marker::PhantomData<CLOCK>,
+}
+
+impl<CLOCK: crate::clock::Clock> PartialEq for Baudrate<CLOCK> {
+    fn eq(&self, other: &Self) -> bool {
+        self.compare_value() == other.compare_value()
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> Eq for Baudrate<CLOCK> {}
+
+impl<CLOCK: crate::clock::Clock> PartialOrd for Baudrate<CLOCK> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.compare_value().cmp(&other.compare_value()))
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> Ord for Baudrate<CLOCK> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        other.compare_value().cmp(&self.compare_value())
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> From<u32> for Baudrate<CLOCK> {
+    fn from(baud: u32) -> Self {
+        Baudrate::new(baud)
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> Baudrate<CLOCK> {
+    pub fn new(baud: u32) -> Baudrate<CLOCK> {
+        let brr = (CLOCK::FREQ / 4 / baud - 1) / 2;
+        debug_assert!(brr <= u16::MAX as u32);
+
+        Baudrate {
+            ubrr: brr as u16,
+            u2x: true,
+            _clock: ::core::marker::PhantomData,
+        }
```

Comment:
Hmm, now you've removed too much:  If the baudrate is very low, the `ubrr` you're calculating here could go too high (maximum allowed value is 4095).  For this situation, you should fall back to non-u2x mode:

```suggestion
        let mut ubrr = (CLOCK::FREQ / 4 / baud - 1) / 2;
        let mut u2x = true;
        debug_assert!(brr <= u16::MAX as u32);
        
        if ubrr > 4095 {
            u2x = false;
            ubrr = (CLOCK::FREQ / 8 / baud - 1) / 2;
        }

        Baudrate {
            ubrr: ubrr as u16,
            u2x: u2x,
            _clock: ::core::marker::PhantomData,
        }
```